src/linux/ptrace_dumper.rs

```diff
@@ -101,28 +152,56 @@ impl PtraceDumper {
             mappings: Vec::new(),
             page_size: 0,
         };
-        dumper.init(stop_timeout)?;
-        Ok(dumper)
+        let soft_errors = dumper.init(stop_timeout)?;
+        Ok((dumper, soft_errors))
     }
 
     // TODO: late_init for chromeos and android
-    pub fn init(&mut self, stop_timeout: Duration) -> Result<(), InitError> {
+    pub fn init(&mut self, stop_timeout: Duration) -> Result<SoftErrorList<InitError>, InitError> {
+        let mut soft_errors = SoftErrorList::default();
+
         // Stopping the process is best-effort.
         if let Err(e) = self.stop_process(stop_timeout) {
-            log::warn!("failed to stop process {}: {e}", self.pid);
+            soft_errors.push(InitError::StopProcessFailed(e));
         }
 
-        if let Err(e) = self.auxv.try_filling_missing_info(self.pid) {
-            log::warn!("failed trying to fill in missing auxv info: {e}");
+        // Even if we completely fail to fill in any additional Auxv info, we can still press
+        // forward.
+        match self.auxv.try_filling_missing_info(self.pid) {
+            Ok(auxv_soft_errors) if !auxv_soft_errors.is_empty() => {
+                soft_errors.push(InitError::FillMissingAuxvInfoErrors(auxv_soft_errors));
+            }
+            Err(e) => {
+                soft_errors.push(InitError::FillMissingAuxvInfoFailed(e));
+            }
+            _ => (),
+        }
+
+        // If we completely fail to enumerate any threads... Some information is still better than
+        // no information!
+        match self.enumerate_threads() {
+            Ok(enumerate_soft_errors) if !enumerate_soft_errors.is_empty() => {
+                soft_errors.push(InitError::EnumerateThreadsErrors(enumerate_soft_errors));
+            }
+            Err(e) => {
+                soft_errors.push(InitError::EnumerateThreadsFailed(Box::new(e)));
+            }
+            _ => (),
+        }
+
+        // Same with mappings -- Some information is still better than no information!
+        match self.enumerate_mappings() {
+            Ok(()) => (),
+            Err(e) => {
+                soft_errors.push(InitError::EnumerateMappingsFailed(Box::new(e)));
+            }
```

Comment:
Looking how many times this pattern is repeated I wonder if it wouldn't be best to put the soft error list into the dumper. Then we could have a method that takes a result and records the error before returning, so we could write more straight-line code like:
```rust
let res: Result<Something, Error> = something_that_might_fail();
let foo = self.record_error(res)?;
// Do stuff with foo
```

Reply:
Totally fair -- I think you will find that it's a lot more concise now :)